### PR TITLE
Use Application.home consistently as opposed to ETSConfig

### LINKED ISSUE #464: Manipulating ETSConfig in `Application. _initialize_application_home` is problematic
In `Application._initialize_application_home`, the `application_home` directory of `ETSConfig` is mutated:
https://github.com/enthought/envisage/blob/6e5c7ba004e8700bd009c3d2b9444d543709a4d7/envisage/application.py#L486-L495

This is good in the sense that it brings the two into alignment if they are different, and it's _definitely_ good that the method makes sure that the directory exists(!), but that allows the possibility of objects created prior to the application having the old value.  The most likely case is for `Preferences` objects to get the wrong thing, in code that creates a custom preferences object, like the (not unreasonable) following:
```
from importlib.resources import as_file, files

with as_file(files(app.mod) / "default.ini) as path:
     preferences = ScopedPreferences()
     preferences.get_scope("default").load(path)

# later
app = Application(preferences=preferences)
```

The implementation of `ScopedPreferences.get_scope` as a side-effect sets the default value of its `application_preferences_filename` trait using the value of `ETSConfig.application_home` _before_ the creation of the `Application` object changes this, so it will potentially target the wrong file (in a directory that may not even exist).

Even worse than the above example, it is possible that even the default `preferences` could end up being created before `_initialize_application_home` is called if, for example, traits listeners are set up on it in the class in a way that causes early instantiation of the preferences from the default (eg. something that listens for changes to the preferences instance to hook up and disconnect preference listeners).

To some extent, this boils down to "do not mess with global state that you do not own."

I'm not sure what the fix should be.

### PR DESCRIPTION
**Pull request by corranwebster**
I think this is the Right Thing, but may cause some breakage of existing code.  Some discussion may be needed.

Fixes #464.

### CODE REVIEW

File: envisage/ui/tasks/tasks_application.py [hunk 1]
```diff
@@ -461,11 +461,8 @@ def _gui_default(self):
         return GUI(splash_screen=self.splash_screen)
 
     def _state_location_default(self):
-        state_location = os.path.join(
-            ETSConfig.application_home, "tasks", ETSConfig.toolkit
-        )
-        if not os.path.exists(state_location):
-            os.makedirs(state_location)
+        state_location = os.path.join(self.home, "tasks", ETSConfig.toolkit)
+        os.makedirs(state_location, mode=0o700, exist_ok=True)
```
Comment (corranwebster) [inline envisage/ui/tasks/tasks_application.py:465]: Possibly should shift this to `_initialize_application_home`?

Comment (mdickinson) [reply envisage/ui/tasks/tasks_application.py:465]: Yes, agreed that extending `_initialize_application_home` seems like the right thing here. I'll make a quick commit to change that.

File: envisage/application.py [hunk 2]
```diff
@@ -155,30 +155,23 @@ def __init__(self, plugins=None, **traits):
 
     def _home_default(self):
         """ Trait initializer. """
-
```
Comment (mdickinson) [inline envisage/application.py:158]: Nit: let's keep the prevailing style in this PR.

**Review by mdickinson — APPROVED** [review]
LGTM; I took the liberty of making some minor style consistency fixes and moving the directory creation to `_initialize_application_home` as you suggested.

**Comment by mdickinson** [comment]
Merging master in to get all the new CI runs.

**Comment by mdickinson** [comment]
Looks like the CI failure was a random EDM installation failure. I've restarted the build.

**Pull request merged**